src/platform/Zephyr/KeyValueStoreManagerImpl.cpp

```diff
@@ -0,0 +1,113 @@
+/*
+ *
+ *    Copyright (c) 2021 Project CHIP Authors
+ *    All rights reserved.
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+/**
+ *    @file
+ *          Platform-specific key value storage implementation for Zephyr
+ */
+
+#include <platform/KeyValueStoreManager.h>
+#include <support/CodeUtils.h>
+#include <support/ReturnMacros.h>
+#include <support/logging/CHIPLogging.h>
+
+#include <logging/log.h>
+#include <settings/settings.h>
+
+namespace chip {
+namespace DeviceLayer {
+namespace PersistedStorage {
+
+KeyValueStoreManagerImpl KeyValueStoreManagerImpl::sInstance;
+
+void KeyValueStoreManagerImpl::Init()
+{
+    VerifyOrDie(settings_subsys_init() == 0);
+}
+
+struct ReadEntry
+{
+    void * destination;           // destination address
+    size_t destinationBufferSize; // size of destination buffer
+    size_t readSize;              // [out] size of read entry value
+    CHIP_ERROR result;            // [out] read result
+};
+
+static int LoadEntryCallback(const char * name, size_t entrySize, settings_read_cb readCb, void * cbArg, void * param)
+{
+    // If requested a key X, process just node X and ignore all its descendants: X/*
+    if (settings_name_next(name, nullptr) > 0)
+        return 0;
```

Comment:
consider setting the result to be CHIP_ERROR_PERSISTED_STORAGE_VALUE_NOT_FOUND  and not rely on defaults in _Get.

Maybe initializing readSize may be useful too.

Reply:
@kkasperczyk-no ?

Reply:
Ok, as you wrote values should be set properly by default, but overriding it here will prevent from potential bugs in the future if somebody unaware of that will make changes in _Get. Done.

Reply:
Actually, it was correct before :)
`settings_load_subtree_direct("some/path", LoadEntryCallback, &entry);` causes `LoadEntryCallback` to be called for "some/path" key and for all descendants of the key, e.g. "some/path/child1", "some/path/child2". This check was to make sure that only "some/path" is processed in the callback. It shouldn't matter in practice as I presume we will use a flat key structure in CHIP, but if we wouldn't, the the current code might override the result of processing the "some/path" node with the result of processing its children.

Reply:
FYI, `settings_name_next` returns a position of the next path component.

Reply:
Ok, so I restored previous version to ensure proper execution in case of nested structure.